### ISSUE TITLE
handling of non-string AMQP headers

### ISSUE THREAD

**Issue by petersilva** [MetPX/sarracenia#323]
an independent implementation is producing messages with non-string AMQP headers for v02 messages.

```

2020-03-26 01:25:05,926 [INFO] post_log notice=20200326012506.053 http://hostlala.gc.ca /20200326/MSC-DMS-DEV/RCM/01/RCM2_OK1055005_PK1066316_1_SCLNB_20200325_164623_HHHV_GRD.tar headers={'sum': 'a,1c76c7987397124a716416d69115c0e5', 'x-delay': 0, 'sundew_extension': 'CISICE:DMS:CMC:RCM_PRODUCT:TAR', 'from_cluster': 'DDSR.CMC-DEV', 'parts': '1,10737418240,1,0,0', 'to_clusters': 'DDI,DDSR', 'source': 'MSC-DMS-DEV'}
2020-03-26 01:25:05,926 [ERROR] sr_sarra/run going badly with 'int' object has no attribute 'encode', sleeping for 5.120.
2020-03-26 01:25:05,926 [DEBUG] Exception details:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/sarra/sr_subscribe.py", line 1759, in run
    ok = self.process_message()
  File "/usr/lib/python3/dist-packages/sarra/sr_subscribe.py", line 942, in process_message
    ok = self.__do_tasks__()
  File "/usr/lib/python3/dist-packages/sarra/sr_subscribe.py", line 421, in __do_tasks__
    if not plugin(self): return False
  File "/usr/lib/python3/dist-packages/sarra/sr_subscribe.py", line 1571, in doit_download
    ok = self.__on_post__()
  File "/usr/lib/python3/dist-packages/sarra/sr_subscribe.py", line 681, in __on_post__
    ok = self.msg.post(self)
  File "/usr/lib/python3/dist-packages/sarra/sr_message.py", line 702, in post
    self.headers[h] = json.dumps( self.headers[h] )
AttributeError: 'int' object has no attribute 'encode'


```
It looks like this other implementation is producing headers with the "integer" type, and
those aren't being digested well by this implementation. hmm...

**Comment by petersilva** [comment]
OK, I take it back... I do not know why it is broken...

```
>>> import json
>>> k={ 'hoho':'1', 'haha':3 }
>>> json.dumps(k)

```
since that works fine, I don't get why this error is being thrown... this was in python 3.8... will try some older ones.

**Comment by petersilva** [comment]
both 2.7 and 3.6 behave the same... OK I mis-interpreted the original problem... hang on.

**Comment by petersilva** [comment]
even json.dumps(k['haha']) works fine in all cases...

**Comment by habilinour** [comment]
The issue is withe the following line in the code which crashes if the header is a list or integer
 if len(self.headers[h].encode("utf8")) >= amqp_ss_maxlen:

**Comment by habilinour** [comment]
As the log shows the error is triggered at the line 702 in sr_message.py
File "/usr/lib/python3/dist-packages/sarra/sr_message.py", line 702, in post

**Comment by petersilva** [comment]
The error is shown in tn the stack trace:
```

  File "/usr/lib/python3/dist-packages/sarra/sr_message.py", line 702, in post
    self.headers[h] = json.dumps( self.headers[h] )
AttributeError: 'int' object has no attribute 'encode'

```

the line doing the json.dumps is the thing generating the error.  The line before that is an if statement testing that  " type(self.headers[h]) is dict  " ... the thing I don't get is why is a header a dict... probably need a debug line right there. just committed that.

Can you try again to subscribe to these messages using the master branch in git?

**Comment by habilinour** [comment]
I have tried the master branch without luck. Currently I'm using a plugin that removes all the non-string headers and it works.

**Comment by petersilva** [comment]
it wasn´t a fix, it was supposed to print a message telling us which header it was going to fail on... can you look at the log lines just before it failed?  (It´s fine to use a shovel that reposts... so that the client stuff can move forward with your work-around, but we keep trying to resolve the undelying issue.)

**Comment by petersilva** [comment]
I wanted to see what the header was that it was failing on.  It was just a guess that it was x-delay... it could be something else.

**Comment by habilinour** [comment]
The new code does not print the debug  message since no header is dict
```
if type(self.headers[h]) is dict:
      self.logger.debug( "dict header flattening to a string: header[ %s ] = %s " % ( h, self.headers[h] ) )

```

**Comment by habilinour** [comment]
Here are the headers in the message. All are str except x-delay which int
```
2020-04-01 23:51:28,732 [DEBUG]  Header from_cluster (value = DDSR.CMC-DEV, type = <class 'str'>)
2020-04-01 23:51:28,732 [DEBUG]  Header to_clusters (value = DDI,DDSR, type = <class 'str'>)
2020-04-01 23:51:28,732 [DEBUG]  Header sum (value = a,6227ff607eadac4d9d12a0d9c8165b0f, type = <class 'str'>)
2020-04-01 23:51:28,732 [DEBUG]  Header x-delay (value = 0, type = <class 'int'>)
2020-04-01 23:51:28,732 [DEBUG]  Header parts (value = 1,10737418240,1,0,0, type = <class 'str'>)
2020-04-01 23:51:28,732 [DEBUG]  Header source (value = MSC-DMS-DEV, type = <class 'str'>)
2020-04-01 23:51:28,732 [DEBUG]  Header rename (value = RCM1_OK1059006_PK1069296_2_SC50MB_20200401_210535_HHHV_GRD.tar, type = <class 'str'>)
2020-04-01 23:51:28,732 [DEBUG]  Header sundew_extension (value = CISISTOP:DMS:CMC:RCM_PRODUCT:TAR, type = <class 'str'>)
```

**Comment by petersilva** [comment]
OK, but line 702... the one that is complaining... is in an if statement ... so it only happens if a header is a dict... ah... you are saying it is actually 703... why would python complain about the wrong line?  ok... it is the next line... like you said in the first place.
 I have a fix for that... try again?

**Comment by habilinour** [comment]
I have run some tests with the release from the master branch and it fixes the issue.

**Comment by petersilva** [comment]
fixed by 2.20.04b2
